hive/agents/deep_dyna_q.py

```diff
@@ -0,0 +1,561 @@
+import copy
+import os
+import pickle
+
+import gymnasium as gym
+import numpy as np
+import torch
+
+from hive.agents.agent import Agent
+from hive.agents.qnets.base import FunctionApproximator
+from hive.agents.qnets.qnet_heads import DQNNetwork
+from hive.agents.world_models.base import WorldModel
+from hive.agents.qnets.utils import (
+    InitializationFn,
+    calculate_output_dim,
+    create_init_weights_fn,
+)
+from hive.replays import BaseReplayBuffer, CircularReplayBuffer
+from hive.utils.loggers import Logger, NullLogger
+from hive.utils.schedule import (
+    LinearSchedule,
+    PeriodicSchedule,
+    Schedule,
+    SwitchSchedule,
+)
+from hive.utils.utils import LossFn, OptimizerFn, create_folder, seeder
+
+
+class DeepDynaQ(Agent):
+    """An agent implementing the nonlinear Dyna-Q algorithm."""
+
+    def __init__(
+        self,
+        observation_space: gym.spaces.Box,
+        action_space: gym.spaces.Discrete,
+        dyna_model: WorldModel,
+        value_representation_net: FunctionApproximator,
+        stack_size: int = 1,
+        id=0,
+        model_optimizer_fn: OptimizerFn = None,
+        value_optimizer_fn: OptimizerFn = None,
+        observation_loss_fn: LossFn = None,
+        observation_normalization_factor: float = 1.0,
+        reward_loss_fn: LossFn = None,
+        terminated_loss_fn: LossFn = None,
+        value_loss_fn: LossFn = None,
+        init_fn: InitializationFn = None,
+        learning_buffer: BaseReplayBuffer = None,
+        planning_buffer: BaseReplayBuffer = None,
+        discount_rate: float = 0.99,
+        n_step: int = 1,
+        grad_clip: float = None,
+        reward_clip: float = None,
+        update_period_schedule: Schedule = None,
+        target_net_soft_update: bool = False,
+        target_net_update_fraction: float = 0.05,
+        target_net_update_schedule: Schedule = None,
+        epsilon_schedule: Schedule = None,
+        test_epsilon: float = 0.001,
+        min_replay_history: int = 5000,
+        learning_batch_size: int = 32,
+        planning_batch_size: int = 32,
+        num_learning_steps: int = 1,
+        num_planning_steps: int = 1,
+        save_agent_info: bool = False,
+        device="cpu",
+        logger: Logger = None,
+        log_frequency: int = 100,
+    ):
+        """
+        Args:
```

Comment:
Please be careful to ensure the docstrings don't exceed 88 characters in length, including preceding whitespace

Reply:
This is applicable to all your functions